File: javascript/173.binary-search-tree-iterator.js
/*
 * @lc app=leetcode id=173 lang=javascript
 *
 * [173] Binary Search Tree Iterator
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 */
var BSTIterator = function(root) {
  this.stack = []
  this.currIndex = 0

  this.partialOrder(root)
};

BSTIterator.prototype.partialOrder = function(root) {
  let partialInOrder = (node) => {
    while (node !== null) {
      this.stack.push(node)
      node = node.left
    }
  }
  partialInOrder(root)
}

/**
 * @return {number}
 */
BSTIterator.prototype.next = function() {
  let top = this.stack.pop()
  if (top) {
    this.partialOrder(top.right)
    return top.val
  }
  return null
};

/**
 * @return {boolean}
 */
BSTIterator.prototype.hasNext = function() {
  return this.stack.length > 0
};

/** 
 * Your BSTIterator object will be instantiated and called as such:
 * var obj = new BSTIterator(root)
 * var param_1 = obj.next()
 * var param_2 = obj.hasNext()
 */
// @lc code=end

File: cpp/173.binary-search-tree-iterator.cpp
/*
 * @lc app=leetcode id=173 lang=cpp
 *
 * [173] Binary Search Tree Iterator
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class BSTIterator
{
public:
    stack<TreeNode *> stacker;
    BSTIterator(TreeNode *root)
    {
        // inorder traversal is left D right
        // traverse partial order 'left'
        // store order in call
        // O(n) time O(n) space
        partialOrder(root);
    }

    void partialOrder(TreeNode *root)
    {
        while (root)
        {
            stacker.push(root);
            root = root->left;
        }
        return;
    }

    int next()
    {
        TreeNode *top = stacker.top();
        stacker.pop();
        if (top)
        {
            partialOrder(top->right);
            return top->val;
        }
        return -1;
    }

    bool hasNext()
    {
        return stacker.size() > 0;
    }
};

/**
 * Your BSTIterator object will be instantiated and called as such:
 * BSTIterator* obj = new BSTIterator(root);
 * int param_1 = obj->next();
 * bool param_2 = obj->hasNext();
 */
// @lc code=end
